base/bundle.hpp

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <experimental/optional>
+#include <functional>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "base/macros.hpp"
+#include "base/not_null.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+// Functions that can be cooperatively aborted should check |AbortRequested()|
+// when appropriate.
+extern thread_local std::function<bool()> AbortRequested;
+
+class Bundle {
+  class Task;
```

Comment:
I'm not in love with the names, it exposes the ugly `TaskHandle` and keeps the nice `Task` internally.  Maybe `Task` and `TaskState`?